### PR TITLE
The extended partition (if any) may not be the last on the disk.

### PR DESCRIPTION
**Pull request by zvin**
Connects to #19

Change-Type: minor

### CODE REVIEW

File: lib/partitioninfo.coffee [hunk 1]
```diff
@@ -46,15 +48,18 @@ getLogicalPartitions = (disk, offset, extendedPartitionOffset = offset, limit =
...
 	.then (buf) ->
 		for p in getPartitionsFromMBRBuf(buf)
 			result.push(partitionDict(p, offset))
-			if MBR.Partition.isExtended(p.type) and getLogical
-				return getLogicalPartitions(disk, p.byteOffset())
-				.then (logicalPartitions) ->
-					result.push(logicalPartitions...)
-					result
+			if MBR.Partition.isExtended(p.type)
+				extended = p
```
Comment (lurch) [inline lib/partitioninfo.coffee:57]: Should this check that `extended` is still `null`, and error otherwise?  (i.e. double-check that there's not two extended partitions)

File: tests/partitioninfo.spec.coffee [hunk 2]
```diff
@@ -69,3 +70,59 @@ describe 'Partitioninfo:', ->
...
+				type: 131
+
+		it 'should return an error when asked for a non existing partition', ->
+			promise = partitioninfo.get(DISK2_PATH, 10)
+			m.chai.expect(promise).to.be.rejectedWith('Partition not found: 10.')
+
+	describe '.getPartitions() disk 2', ->
+		it 'should list all partitions of a disk image', ->
+			promise = partitioninfo.getPartitions(DISK2_PATH)
+			m.chai.expect(promise).to.eventually.become([
```
Comment (lurch) [inline tests/partitioninfo.spec.coffee:142]: Hmmm, with this `getPartitions` function returning an array of 6 partitions, I wonder if people might expect (from an API point of view) that `partitioninfo.get(DISK2_PATH, 4)` would return the 4th item in this array, when it would actually result in `partitionNotFoundError(4)` ?  (i.e. `partitioninfo.get(DISK2_PATH, n)` doesn't necessarily equal `partitioninfo.getPartitions(DISK2_PATH)[n - 1]` )

(I also wonder if `partitioninfo.get(DISK2_PATH, 4)` would be a good unit-test to add in any case?)

File: tests/partitioninfo.spec.coffee [hunk 3]
```diff
@@ -69,3 +70,59 @@ describe 'Partitioninfo:', ->
...
+				{
+					'offset': 5242880
+					'size': 1048576
+					'type': 131
+				}
+				{
+					'offset': 7340032
+					'size': 1048576
+					'type': 131
+				}
```
Comment (lurch) [inline tests/partitioninfo.spec.coffee:172]: With the results presented like this, it's not immediately obvious (without "doing the math" to work out which partitions overlap) whether this disk2.img is actually structured as:
```
/dev/sda1 /dev/sda2 [ /dev/sda5 /dev/sda6 /dev/sda7 /dev/sda8 ]

/dev/sda1 /dev/sda2 [ /dev/sda5 /dev/sda6 /dev/sda7 ] /dev/sda3

/dev/sda1 /dev/sda2 [ /dev/sda5 /dev/sda6 ] /dev/sda3 /dev/sda4
```

**Review by lurch — COMMENTED** [review]
Took me a while to figure out how this code is actually working, but yeah it looks more sensible now. But as I'm not a CoffeeScript programmer, IMHO it should be reviewed by somebody who is ;-)

File: tests/partitioninfo.spec.coffee (hunk 3 above)
Comment (lurch) [reply tests/partitioninfo.spec.coffee:172]: ...and it's also not immediately obvious (if it matters?) that `.getPartitions()` returns them in the order `[/dev/sda1, /dev/sda2, /dev/sda3, /dev/sda5, /dev/sda6, /dev/sda7]`.

Would it also be useful for `.getPartitions()` to have an `ignoreExtended` parameter (defaulting to false), which if set would then return just   `[/dev/sda1, /dev/sda3, /dev/sda5, /dev/sda6, /dev/sda7]`?

(in case it's not obvious, I'm  only using Linux device-names here as an unambiguous short-hand ;-)  I'm not suggesting that this module should return Linux device-names)

**Comment by lurch** [comment]
How come #17 modifies `build/partitioninfo.js` but this PR doesn't?

**Comment by zvin** [comment]
@lurch 
About `build/*.js`: good catch, that's because I didn't run `npm install` that generates the js files.

File: tests/partitioninfo.spec.coffee (hunk 2 above)
Comment (zvin) [reply tests/partitioninfo.spec.coffee:142]: I didn't want to break the API so `getPartitions()` is still returning an array.
I'll add a note in the README that `getPartitions()[n - 1]` may not be equal to `get(n)`.
The other solution would be to make `getPartitions()` return an object.

I'll add the unit-test.

File: tests/partitioninfo.spec.coffee (hunk 3 above)
Comment (zvin) [reply tests/partitioninfo.spec.coffee:172]: I'll add a note in the README about the partitions order.
`ignoreExtended` is a good idea, I'll add that.

File: lib/partitioninfo.coffee (hunk 1 above)
Comment (zvin) [reply lib/partitioninfo.coffee:57]: I'm not sure this is required. I don't know any tool that would create an image with more than one extended partition for testing.

File: tests/partitioninfo.spec.coffee (hunk 3 above)
Comment (zvin) [reply tests/partitioninfo.spec.coffee:172]: I used `includeExtended=true` instead of `ignoreExtended=false`

File: tests/partitioninfo.spec.coffee (hunk 2 above)
Comment (zvin) [reply tests/partitioninfo.spec.coffee:142]: added

File: tests/partitioninfo.spec.coffee [hunk 4]
```diff
@@ -69,3 +70,135 @@ describe 'Partitioninfo:', ->
...
+			])
+
+	describe '.getPartitions() disk 2 includeExtended = false', ->
+		it 'should list all partitions of a disk image', ->
```
Comment (lurch) [inline tests/partitioninfo.spec.coffee]: Minor nit, but I wonder if the 'it' line here should be different to the one on line 140?

**Comment by lurch** [comment]
Great to see updated documentation :+1:

File: tests/partitioninfo.spec.coffee (hunk 4 above)
Comment (zvin) [reply tests/partitioninfo.spec.coffee]: Right, I just updated it.

File: lib/partitioninfo.coffee [hunk 5]
```diff
@@ -43,18 +45,24 @@ getLogicalPartitions = (disk, offset, extendedPartitionOffset = offset, limit =
 					result
 		result
 
-getPartitions = (disk, offset, getLogical = true) ->
+getPartitions = (disk, includeExtended, offset, getLogical = true) ->
```
Comment (jviotti) [inline lib/partitioninfo.coffee]: Can we create an `options` object containing both `getLogical` and `includeExtended`? Calling the function with a couple of boolean values was a bit hard to follow

Comment (zvin) [reply lib/partitioninfo.coffee]: @jviotti updated, the prototype is now (disk, options).
options is { includeExtended, offset, getLogical }

File: lib/partitioninfo.coffee [hunk 6]
```diff
@@ -92,6 +102,21 @@ callWithDisk = (fn, pathOrDisk, args...) ->
 
 ###*
 # @summary Get information from a partition
+#
+# @description `getPartitions()` returns an Array.
```
Comment (lurch) [inline lib/partitioninfo.coffee]: Looks like some parts of the `@description` here should be in the jsdoc for `exports.getPartitions`, rather than `exports.get` as it is now?

Comment (zvin) [reply lib/partitioninfo.coffee]: moved

Comment (lurch) [reply lib/partitioninfo.coffee]: Looks like you've forgotten to update the auto-generated files? ;-)

Comment (zvin) [reply lib/partitioninfo.coffee]: Good catch, files updated.
